File: leetcode/160.py
# 160. Intersection of Two Linked Lists

# Approach 1. Nexted while loop 
# Time Complexity: O(n*m) (worst), Space Complexity: O(1)
# class Solution:
#     def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
#         if not headA or not headB:
#             return null
        
#         tempheadB = headB 
#         while headA:
#             while headB:
#                 if headA == headB:
#                     return headA
#                 headB = headB.next
#             headA = headA.next
#             headB = tempheadB

# Approach 2. set
# Time Complexity: O(n+m), Space Complexity: O(m) or O(n)
class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        setB = set()

        while headB is not None:
            setB.add(headB)
            headB = headB.next

        while headA is not None:
            if headA in setB:
                return headA
            headA = headA.next

        return None

# Approach 3. two pointers
# Time Complexity: O(2(n+m)) ~= O(n+m), Space Complexity: O(1)
class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        pA = headA
        pB = headB

        while pA != pB:
            pA = headB if pA is None else pA.next
            pB = headA if pB is None else pB.next
        
        return pA

File: leetcode/160.cpp
// 160. Intersection of Two Linked Lists

#include <iostream>
#include <set>

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

// 1. Brute And Force: TC: O(n*m), SC: O(1)
// class Solution {
// public:
//     ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
//         ListNode* currA = headA;
//         while (currA) {
//             ListNode* currB = headB;
//             while (currB) {
//                 if (currA == currB) {
//                     return currA;
//                 }
//                 currB = currB->next;
//             }
//             currA = currA->next;
//         }
//         return NULL;
//     }
// };

// 2. My solution: TC: O(n)
// class Solution {
// public:
//     ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
//         ListNode* currA = headA;
//         ListNode* currB = headB;
//         while (currA  && currB) {
//             if (currA == currB)
//                 return currA;
//             currA = currA->next;
//             currB = currB->next;
//             if (!currA && !currB)
//                 return NULL;
//             if (!currA) 
//                 currA = headB;
//             if (!currB)
//                 currB = headA;
//         }
//         return NULL;
//     }
// };

// 3. 
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        set<ListNode *> nodes_in_B;

        while (headB != nullptr) {        
            nodes_in_B.insert(headB);
            headB = headB->next;
        }

        while (headA != nullptr) {

            if (nodes_in_B.find(headA) != nodes_in_B.end()) {
                return headA;
            }
            headA = headA->next;
        }

        return nullptr;
    }
};